frontend/src/views/Game.tsx

```diff
@@ -1,45 +1,65 @@
 import React, { useEffect, useState } from 'react';
-import { useLocation } from 'react-router-dom';
+import { useHistory, useLocation } from 'react-router-dom';
 import SplitterLayout from 'react-splitter-layout';
+import { useBeforeunload } from 'react-beforeunload';
 import Editor from '../components/core/Editor';
-import { ErrorResponse } from '../api/Error';
+import { errorHandler, ErrorResponse } from '../api/Error';
 import { Problem, getProblems } from '../api/Problem';
 import { Room } from '../api/Room';
-import {
+import MainContainer, {
   FlexContainer, FlexInfoBar, Panel, SplitterContainer,
 } from '../components/core/Container';
 import ErrorMessage from '../components/core/Error';
 import { ProblemHeaderText, Text } from '../components/core/Text';
 import 'react-splitter-layout/lib/index.css';
 import { checkLocationState } from '../util/Utility';
+import Loading from '../components/core/Loading';
 
 type LocationState = {
   room: Room,
 }
 
 function GamePage() {
+  // Get history object to be able to move between different pages
+  const history = useHistory();
   const location = useLocation<LocationState>();
   const [room, setRoom] = useState<Room | null>(null);
   const [problems, setProblems] = useState<Problem[]>([]);
+  const [fullPageLoading, setFullPageLoading] = useState<boolean>(true);
   const [error, setError] = useState<string>('');
 
+  /**
+   * Display beforeUnload message to inform the user that they may lose
+   * their code / data if they leave the page.
+   * Some browsers will display this message, others will display a fixed
+   * message; see https://github.com/jacobbuck/react-beforeunload.
+   */
+  useBeforeunload(() => 'Leaving this page may cause you to lose your current code and data.');
```

Comment:
The fact that unload isn't capitalized is bothering me way more than it should haha

Reply:
Here's what we do -- make a new npm package that just uses this existing package but then renames it, publish the package, then use _that_ in our codebase. Or we can make a Utility function that just invokes this method but is called `useBeforeUnload`, then we use _that_ function.

For now though, I'll leave it as-is. 🙂